src/azure-cli-core/azure/cli/core/commands/__init__.py

```diff
@@ -142,14 +143,25 @@ def wrapped(func):
 
 class CliCommand(object):
 
-    def __init__(self, name, handler, description=None, table_transformer=None):
+    def __init__(self, name, handler, description=None, table_transformer=None,
+                 arguments_loader=None, description_loader=None):
         self.name = name
         self.handler = handler
-        self.description = description
         self.help = None
+        self.description = description_loader() \
+                           if description_loader and self._should_load_description() \
+                           else description
         self.arguments = {}
+        self.arguments_loader = arguments_loader
         self.table_transformer = table_transformer
 
+    def load_arguments(self):
+        if self.arguments_loader:
+            self.arguments.update(self.arguments_loader())
+
+    def _should_load_description(self):
+        return not APPLICATION.session['completer_active']
```

Comment:
When help is requested, it's too late to dynamically try and load the description as the command table has already been passed to the parser.

With this approach, the description is only loaded when a command is actually executing so the overhead (about 0.1/0.2 secs) is small in comparison to whatever the command is doing.

With tab completion, we don't need the description so can remove this overhead.

That was the thinking.